#VRML_SIM R2020b utf8
WorldInfo {
  info [
    "Model of the Salamandra Robotica amphibious robot developped at the BIOROB (http://biorob.epfl.ch). This demo exemplifies how Webots automotically generates hydrostatic and hydrodynamic forces."
    "Authors: Jerome Braure (original work), Yvan Bourquin (this demo) and Luc Guyot (adaptation to the generic Webots fluid dynamics features)"
    "Biorobotics Laboratory, EPFL, Lausanne"
    "Ecole Polytechnique Federale de Lausanne (EPFL), Switzerland"
    ""
    "Related publication:"
    "A. Ijspeert, A. Crespi, D. Ryczko, and J.M. Cabelguen. From swimming to walking with a salamander robot driven by a spinal cord model. Science, 315(5817):1416-1420, 2007."
  ]
  title "Salamandra Robotica"
  basicTimeStep 4
  coordinateSystem "NUE"
  contactProperties [
    ContactProperties {
      material1 "leg"
      material2 "floor"
      coulombFriction [
        5
      ]
    }
    ContactProperties {
      material1 "body"
      material2 "floor"
      coulombFriction [
        5
      ]
    }
  ]
}
Viewpoint {
  orientation 0.9825992334999789 -0.18414542152859883 0.024273649402356236 5.587600481634637
  position 4.123160411300494 1.6528206100523637 5.037126445274621
}
TexturedBackground {
}
TexturedBackgroundLight {
}
DEF FLOOR Solid {
  children [
    DEF GND Group {
      children [
        Transform {
          translation -4.42876 -0.25 0
          children [
            Shape {
              appearance DEF GROUND_APPEARANCE PBRAppearance {
                baseColorMap ImageTexture {
                  url [
                    "textures/texture_grid.png"
                  ]
                }
                roughness 1
                metalness 0
              }
              geometry Box {
                size 3.14249 0.001 8
              }
            }
          ]
        }
        Transform {
          rotation 0 0 1 0.087266
          children [
            Shape {
              appearance USE GROUND_APPEARANCE
              geometry Box {
                size 5.73686 0.001 8
              }
            }
          ]
        }
        Transform {
          translation 4.42876 0.25 0
          children [
            Shape {
              appearance USE GROUND_APPEARANCE
              geometry Box {
                size 3.14249 0.001 8
              }
            }
          ]
        }
      ]
    }
  ]
  name "ground"
  contactMaterial "floor"
  boundingObject USE GND
  locked TRUE
}
DEF BIG_ISLAND Solid {
  translation 1.5 0 1.5
  rotation 0 1 0 0.7854
  children [
    DEF BIG_ISLAND_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.6 1 0.6
        roughness 1
        metalness 0
      }
      geometry Box {
        size 1.4 1.4 1.4
      }
    }
  ]
  name "big island"
  boundingObject USE BIG_ISLAND_SHAPE
  locked TRUE
}
DEF SMALL_ISLAND Solid {
  translation -1.5 0 -1.5
  rotation 0 1 0 0.7854
  children [
    DEF SMALL_ISLAND_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.6 1 0.6
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.8 0.8 0.8
      }
    }
  ]
  name "small island"
  boundingObject USE SMALL_ISLAND_SHAPE
  locked TRUE
}
DEF AQUARIUM Solid {
  translation 0 0.1 0
  children [
    DEF SIDES_GROUP Group {
      children [
        Transform {
          translation 0 0 4
          children [
            DEF GLASS_SHAPE1 Shape {
              appearance DEF GLASS_APPEARANCE PBRAppearance {
                baseColor 1 0.6 0.6
                roughness 1
                metalness 0
              }
              geometry Box {
                size 12 0.7 0.02
              }
            }
          ]
        }
        Transform {
          translation 0 0 -4
          children [
            USE GLASS_SHAPE1
          ]
        }
        Transform {
          translation -6 0 0
          rotation 0 1 0 1.5708
          children [
            DEF GLASS_SHAPE2 Shape {
              appearance USE GLASS_APPEARANCE
              geometry Box {
                size 8 0.7 0.02
              }
            }
          ]
        }
        Transform {
          translation 6 0 0
          rotation 0 1 0 1.5708
          children [
            USE GLASS_SHAPE2
          ]
        }
      ]
    }
  ]
  name "aquarium"
  boundingObject USE SIDES_GROUP
  locked TRUE
}
DEF WATER Fluid {
  translation 0 -0.125 0
  children [
    DEF WATER_BOX Shape {
      appearance PBRAppearance {
        baseColor 0.7254901960784313 0.8705882352941177 1
        transparency 0.7
        roughness 0.2
        metalness 0
      }
      geometry Box {
        size 12 0.25 8
      }
    }
  ]
  boundingObject USE WATER_BOX
  locked TRUE
}
DEF Salamander Robot {
  translation 3.4 0.29 3.15
  rotation 0 1 0 1.57
  children [
    Group {
      children [
        Slot {
        }
      ]
    }
    GPS {
    }
    DEF SMALL_SEGMENT Group {
      children [
        Transform {
          children [
            Shape {
              appearance DEF SALAMANDER_COLOR PBRAppearance {
                baseColor 1 0.8 0.3
                roughness 0.3
                metalness 0
              }
              geometry Cylinder {
                height 0.0561
                radius 0.019
              }
            }
          ]
        }
        Transform {
          translation 0 0 0.038
          children [
            Shape {
              appearance USE SALAMANDER_COLOR
              geometry Box {
                size 0.038 0.056 0.076
              }
            }
          ]
        }
      ]
    }
    DEF LEFT_EYE DistanceSensor {
      translation -0.02 0.03 0
      rotation 0 1 0 2.356
      children [
        DEF EYE_SHAPE Shape {
          appearance DEF EYE_APPEARANCE PBRAppearance {
            baseColor 0.2 0.2 0.2
            roughness 0.2
            metalness 0
          }
          geometry Sphere {
            radius 0.005
            subdivision 2
          }
        }
      ]
      name "ds_left"
      lookupTable [
        0 0 0
        0.7 0.7 0
      ]
      numberOfRays 6
      aperture 0.8
    }
    DEF RIGHT_EYE DistanceSensor {
      translation 0.02 0.03 0
      rotation 0 1 0 0.785
      children [
        USE EYE_SHAPE
      ]
      name "ds_right"
      lookupTable [
        0 0 0
        0.7 0.7 0
      ]
      numberOfRays 6
      aperture 0.8
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0 0 0.108
        minStop -0.0001
        maxStop 0.0001
      }
      endPoint DEF SOLID_THORAX Solid {
        translation 0 0 0.108
        rotation 0 1 0 0.00011224817144041612
        children [
          DEF BIG_SEGMENT Group {
            children [
              Transform {
                children [
                  Shape {
                    appearance USE SALAMANDER_COLOR
                    geometry Cylinder {
                      height 0.0561
                      radius 0.032
                    }
                  }
                ]
              }
              Transform {
                translation 0 0 0.0235
                children [
                  Shape {
                    appearance USE SALAMANDER_COLOR
                    geometry Box {
                      size 0.064 0.056 0.047
                    }
                  }
                ]
              }
            ]
          }
          HingeJoint {
            jointParameters HingeJointParameters {
              axis 0 1 0
              anchor 0 0 0.066
            }
            device [
              RotationalMotor {
                name "motor_1"
                minPosition -1.13446
                maxPosition 1.13446
                maxTorque 100
              }
            ]
            endPoint DEF SEGMENT_1 Solid {
              translation 0 0 0.066
              children [
                USE SMALL_SEGMENT
                HingeJoint {
                  jointParameters HingeJointParameters {
                    axis 0 1 0
                    anchor 0 0 0.095
                  }
                  device [
                    RotationalMotor {
                      name "motor_2"
                      minPosition -1.13446
                      maxPosition 1.13446
                      maxTorque 100
                    }
                  ]
                  endPoint DEF SEGMENT_2 Solid {
                    translation 0 0 0.095
                    children [
                      USE SMALL_SEGMENT
                      HingeJoint {
                        jointParameters HingeJointParameters {
                          axis 0 1 0
                          anchor 0 0 0.095
                        }
                        device [
                          RotationalMotor {
                            name "motor_3"
                            minPosition -1.13446
                            maxPosition 1.13446
                            maxTorque 100
                          }
                        ]
                        endPoint DEF SEGMENT_3 Solid {
                          translation 0 0 0.095
                          children [
                            USE SMALL_SEGMENT
                            HingeJoint {
                              jointParameters HingeJointParameters {
                                axis 0 1 0
                                anchor 0 0 0.108
                                minStop -0.0001
                                maxStop 0.0001
                              }
                              endPoint DEF SEGMENT_PELVIS Solid {
                                translation 0 0 0.108
                                children [
                                  USE BIG_SEGMENT
                                  HingeJoint {
                                    jointParameters HingeJointParameters {
                                      axis 0 1 0
                                      anchor 0 0 0.066
                                    }
                                    device [
                                      RotationalMotor {
                                        name "motor_4"
                                        minPosition -1.13446
                                        maxPosition 1.13446
                                        maxTorque 100
                                      }
                                    ]
                                    endPoint DEF SEGMENT_4 Solid {
                                      translation 0 0 0.066
                                      children [
                                        USE SMALL_SEGMENT
                                        HingeJoint {
                                          jointParameters HingeJointParameters {
                                            axis 0 1 0
                                            anchor 0 0 0.095
                                          }
                                          device [
                                            RotationalMotor {
                                              name "motor_5"
                                              minPosition -1.13446
                                              maxPosition 1.13446
                                              maxTorque 100
                                            }
                                          ]
                                          endPoint DEF SEGMENT_5 Solid {
                                            translation 0 0 0.095
                                            children [
                                              USE SMALL_SEGMENT
                                              HingeJoint {
                                                jointParameters HingeJointParameters {
                                                  axis 0 1 0
                                                  anchor 0 0 0.095
                                                }
                                                device [
                                                  RotationalMotor {
                                                    name "motor_6"
                                                    minPosition -1.13466
                                                    maxPosition 1.13446
                                                    maxTorque 100
                                                  }
                                                ]
                                                endPoint DEF SEGMENT_6 Solid {
                                                  translation 0 0 0.095
                                                  children [
                                                    USE SMALL_SEGMENT
                                                  ]
                                                  contactMaterial "body"
                                                  immersionProperties [
                                                    DEF IMMERSION_PROP_SEGMENT ImmersionProperties {
                                                      fluidName "fluid"
                                                      referenceArea "xyz-projection"
                                                      dragForceCoefficients 3.1 1.3 0
                                                      viscousResistanceTorqueCoefficient 5
                                                    }
                                                  ]
                                                  boundingObject USE SMALL_SEGMENT
                                                  physics DEF SEGMENT_PHYSICS Physics {
                                                    density -1
                                                    mass 0.180281
                                                    centerOfMass [
                                                      0 -0.014 0.038
                                                    ]
                                                    inertiaMatrix [
                                                      0.000222864 0.000160531 0.000102659
                                                      0 0 2.70783e-05
                                                    ]
                                                  }
                                                }
                                              }
                                            ]
                                            contactMaterial "body"
                                            immersionProperties [
                                              USE IMMERSION_PROP_SEGMENT
                                            ]
                                            boundingObject USE SMALL_SEGMENT
                                            physics USE SEGMENT_PHYSICS
                                          }
                                        }
                                      ]
                                      contactMaterial "body"
                                      immersionProperties [
                                        USE IMMERSION_PROP_SEGMENT
                                      ]
                                      boundingObject USE SMALL_SEGMENT
                                      physics USE SEGMENT_PHYSICS
                                    }
                                  }
                                ]
                                contactMaterial "body"
                                immersionProperties [
                                  DEF IMMERSION_PROP_BIG_SEGMENT ImmersionProperties {
                                    fluidName "fluid"
                                    referenceArea "xyz-projection"
                                    dragForceCoefficients 1.8 0.95 0.75
                                    viscousResistanceTorqueCoefficient 5
                                  }
                                ]
                                boundingObject USE BIG_SEGMENT
                                physics DEF BIG_SEGMENT_PHYSICS Physics {
                                  density -1
                                  mass 0.279137
                                  centerOfMass [
                                    0 -0.014 0.0235
                                  ]
                                  inertiaMatrix [
                                    0.000269294 0.000224459 0.000210752
                                    0 0 4.75005e-05
                                  ]
                                }
                              }
                            }
                          ]
                          contactMaterial "body"
                          immersionProperties [
                            USE IMMERSION_PROP_SEGMENT
                          ]
                          boundingObject USE SMALL_SEGMENT
                          physics USE SEGMENT_PHYSICS
                        }
                      }
                    ]
                    contactMaterial "body"
                    immersionProperties [
                      USE IMMERSION_PROP_SEGMENT
                    ]
                    boundingObject USE SMALL_SEGMENT
                    physics USE SEGMENT_PHYSICS
                  }
                }
              ]
              contactMaterial "body"
              immersionProperties [
                USE IMMERSION_PROP_SEGMENT
              ]
              boundingObject USE SMALL_SEGMENT
              physics USE SEGMENT_PHYSICS
            }
          }
        ]
        contactMaterial "body"
        immersionProperties [
          USE IMMERSION_PROP_BIG_SEGMENT
        ]
        boundingObject USE BIG_SEGMENT
        physics USE BIG_SEGMENT_PHYSICS
      }
    }
  ]
  name "Salamander1"
  contactMaterial "body"
  immersionProperties [
    DEF IMMERSION_PROP_HEAD ImmersionProperties {
      fluidName "fluid"
      referenceArea "xyz-projection"
      dragForceCoefficients 2 1 1
      viscousResistanceTorqueCoefficient 5
    }
  ]
  boundingObject USE SMALL_SEGMENT
  physics USE SEGMENT_PHYSICS
  controller "salamanderc"
}
